### PR TITLE
Use python for ENV extraction

### PR DESCRIPTION
**Pull request by steelbrain**
We still fallback to `env` when python is not available

### CODE REVIEW

File: src/helpers.js [hunk 1]
```diff
@@ -17,10 +18,14 @@ export const assign = Object.assign || function (target, source) {
   return target
 }
 
+export function splitEnv(contents: string) {
+  return contents.split(contents.includes('\0') ? '\0' : '\n')
```
Comment (Arcanemagus) [inline src/helpers.js:22]: The entire problem is that the previous code was splitting on `\n`, What reason could there be for keeping that broken behavior?

File: src/helpers.js [hunk 2]
```diff
@@ -87,18 +93,34 @@ export function applySugar(environment: Object) {
 }
 
 export function getCommand(): { command: string, options: Object, parameters: Array<string> } {
+  const pythonFile = Path.join(Path.dirname(__dirname), 'index.py')
```
Comment (Arcanemagus) [inline src/helpers.js]: Just do `__dirname, '..'`, no sense in doing this silly "dirname of a dirname" code.

File: src/helpers.js [hunk 3]
```diff
@@ -87,18 +93,34 @@ export function applySugar(environment: Object) {
...
+    stdin.push('source ~/.bashrc 2>/dev/null')
+    stdin.push('source ~/.bash_profile 2>/dev/null')
   } else if (shell === 'zsh') {
-    parameters = ['-c', 'source ~/.zshrc;env;exit']
+    stdin.push('source ~/.zshrc 2>/dev/null')
   } else if (shell === 'fish') {
-    parameters = ['-c', 'source ~/.config/fish/config.fish;env;exit']
+    stdin.push('source ~/.config/fish/config.fish 2>/dev/null')
   }
+  stdin.push(shellScript)
```
Comment (Unrud) [inline src/helpers.js:121]: This doesn't work. The shell script is not compatible with many shells (like **fish** or **csh**).

Comment (steelbrain) [reply src/helpers.js:121]: Thanks for the heads up, I'll test this manually in csh and fish to confirm!

File: src/helpers.js (hunk 2 above)
Comment (Unrud) [inline src/helpers.js]: Just run something like ``python -c 'import os;print("\0".join(map("=".join, dict(os.environ).items())))'``, no need to create a file.

**Review by Arcanemagus — CHANGES_REQUESTED** [review]
I very much prefer the pure shell method in #38 instead of bringing in a dependency on `python`.

File: src/helpers.js (hunk 2 above)
Comment (steelbrain) [reply src/helpers.js]: Removed in latest push

File: src/helpers.js (hunk 1 above)
Comment (steelbrain) [reply src/helpers.js:22]: We are using python splitting when it's available but when it's not we fallback to env, and we still need to split the env output with `\n`

**Comment by steelbrain** [comment]
@Arcanemagus Python is available in all linux distros that I'm aware of and macOS by default. That's why we're using it, even if it doesn't, we still fallback to the `env` way.

The main concern I had with the pure shell method was that it was still getting splitting the output by `\n`, it was just using some regexp to validate, this solution is rock-solid

I am open to suggestions tho :)

File: src/helpers.js (hunk 3 above)
Comment (Unrud) [reply src/helpers.js:121]: You can wrap the script in sh, as I did in #36 for incompatible shells.

File: src/helpers.js [hunk 4]
```diff
@@ -86,19 +92,35 @@ export function applySugar(environment: Object) {
...
 export function getCommand(): { command: string, options: Object, parameters: Array<string> } {
+  const shellScript = coolTrim`
+    if [ "$(which python)" != "" ]; then
+      python -c '${pythonCode}'
+    elif [ "$(which python3)" != "" ]; then
+      python3 -c '${pythonCode}'
+    elif [ "$(which python2.7)" != "" ]; then
+      python2.7 -c '${pythonCode}'
+    else
+      env
```
Comment (Unrud) [inline src/helpers.js:105]: I would remove this. If you really need a fallback, use something like the script from #36.

**Comment by steelbrain** [comment]
@Unrud This was a PoC PR. I will have some time to finish this tonight, but if you want to open a new PR with similar functionality, that would be 👍

**Comment by Unrud** [comment]
> The main concern I had with the pure shell method was that it was still getting splitting the output by ``\n``.

It only does this to find potential names of environment variables. It does further checks to validate that the name is real and it uses a safe method to retrieve the contents of environment variable.

> PR with similar functionality

#40

**Comment by steelbrain** [comment]
Closing in favor of #40

**Pull request closed without merging**